### PR TITLE
Add BlockStmt

### PR DESCRIPTION
**Pull request by tehrengruber**
During development of if-statements #1079 a new FOAST node `BlockStmt` was introduced. The node is used to group multiple `foast.Stmt`s into a single node (e.g. each branch of a statement is a single `BlockStmt`). Primarily to avoid #1079 to rot before the conceptual problems on ITIR are solved these changes were factored out into this PR. However the changes are meaningful on their own as they introduce a cleaner lowering mechanism from the somewhat-stateful FOAST to the functional ITIR.

Additional changes in this PR:
- Fixed small bug in `test_closure_symbols`
- Add `if_` helper function to `itir_makers.py`

### CODE REVIEW

File: src/functional/ffront/foast_pretty_printer.py [hunk 1]
```diff
@@ -11,6 +11,8 @@
 
 PropertyIdentifier: TypeAlias = Union[type[foast.LocatedNode], tuple[type[foast.LocatedNode], str]]
 
+INDENTATION_PREFIX: Final[str] = "  "
```
Comment (egparedes) [inline src/functional/ffront/foast_pretty_printer.py:14]: Is this ever used? And if it's not, why is not `flake8` complaining?

File: src/functional/ffront/foast_to_itir.py [hunk 2]
```diff
@@ -179,6 +180,10 @@ def visit_FieldOperator(self, node: foast.FieldOperator, **kwargs) -> itir.Funct
             if isinstance(node.definition.params[i].type, ts.ScalarType):
                 new_body = im.let(param.id, im.deref_(param.id))(new_body)
 
+        assert isinstance(node.type, ts.FieldOperatorType)
```
Comment (egparedes) [inline src/functional/ffront/foast_to_itir.py:183]: Could you explain a bit what's going on here?

File: src/functional/ffront/itir_makers.py [hunk 3]
```diff
@@ -223,6 +223,11 @@ def tuple_get_(tuple_expr, index):
     return call_("tuple_get")(tuple_expr, index)
 
 
+def if_(cond, true_val, false_val):
```
Comment (egparedes) [inline src/functional/ffront/itir_makers.py:226]: I don't think this change belongs to this PR, does it?

File: tests/functional_tests/ffront_tests/test_foast_parsing.py [hunk 4]
```diff
@@ -350,22 +350,25 @@ def operator_with_refs(inp: Field[..., "float64"], inp2: Field[..., "float32"]):
         return a, b
 
     parsed = FieldOperatorParser.apply_to_function(operator_with_refs)
-    assert "nonlocals_unused" not in parsed.annex.symtable
-    assert "nonlocals" not in parsed.annex.symtable
+    assert "nonlocals_unreferenced" not in {**parsed.annex.symtable, **parsed.body.annex.symtable}
+    assert "nonlocals" not in {**parsed.annex.symtable, **parsed.body.annex.symtable}
```
Comment (egparedes) [inline tests/functional_tests/ffront_tests/test_foast_parsing.py:354]: Is this related to this PR or to something else?

**Review by egparedes — COMMENTED** [review]
It looks good in general. I have some questions for my own understanding and some minor comments. Additionally, it seems there are some changes to improve the quality of the code (for example in the tests) which are not strictly related to `BlockStmt`, so please make sure those changes are also mentioned in the commit message.

File: src/functional/ffront/foast_introspection.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+from functional.ffront import field_operator_ast as foast
+
+
+class StmtReturnKind(enum.IntEnum):
+    UNCONDITIONAL_RETURN = 0
+    NO_RETURN = 2
+
+
+def deduce_stmt_return_kind(node: foast.Stmt) -> StmtReturnKind:
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/functional/ffront/foast_introspection.py:24]: As discussed, but restated here for the record: This function is meant to accept every `foast.Stmt` (so that callers can rely on this) and should be extended for new subclasses.

File: src/functional/ffront/foast_passes/iterable_unpack.py [hunk 6]
```diff
@@ -57,18 +43,16 @@ def _unique_tuple_symbol(self, node: foast.TupleTargetAssign) -> foast.Symbol[An
         self.unique_tuple_symbol_id += 1
         return sym
 
-    def _unroll_tuple_target_assign(
-        self, body: list[foast.LocatedNode]
-    ) -> list[foast.Assign | foast.LocatedNode]:
+    def visit_BlockStmt(self, node: foast.BlockStmt, **kwargs) -> foast.BlockStmt:
         unrolled: list[foast.Assign | foast.LocatedNode] = []
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/functional/ffront/foast_passes/iterable_unpack.py]: Renamed to `unrolled_stmts` and changed the annotation to `list[foast.Assign | foast.BlockStmt | foast.Return]` to signify that the resulting stmts are not of `foast.TupleTargetAssign`.

File: src/functional/ffront/foast_pretty_printer.py [hunk 7]
```diff
@@ -139,11 +142,13 @@ def visit_TernaryExpr(self, node: foast.TernaryExpr, **kwargs) -> str:
 
     Return = as_fmt("return {value}")
 
+    BlockStmt = as_mako("${'\\n'.join(stmts)}")
```
Comment (egparedes) [inline src/functional/ffront/foast_pretty_printer.py:145]: Feel free to ignore, but I just realized that mako is not needed at all here.
```suggestion
    BlockStmt = as_fmt("{'\n'.join(stmts)}")
```

**Review by egparedes — APPROVED** [review]
It looks good. I left a comment with an optional suggestion. Please, remember to mention in the commit message all the different changes added in this PR.

File: src/functional/ffront/foast_pretty_printer.py (hunk 7 above)
Comment (tehrengruber) [reply src/functional/ffront/foast_pretty_printer.py:145]: Format strings sadly don't support newline characters inside braces: https://stackoverflow.com/questions/44780357/how-to-use-newline-n-in-f-string-to-format-output-in-python-3-6
